libr/asm/arch/8051/8051_ass.c

```diff
@@ -958,97 +942,204 @@ static bool mnem_mov_c(char const*const*arg, ut16 pc, ut8**out) {
+		singlearg_immediate (0x77, arg[1], out);
+		return true;
+	} else if (!r_str_casecmp (arg[0], "r0") && resolve_immediate (arg[1] + 1, &src_imm)) {
+		singlearg_immediate (0x78, arg[1], out);
+		return true;
+	} else if (!r_str_casecmp (arg[0], "r1") && resolve_immediate (arg[1] + 1, &src_imm)) {
+		singlearg_immediate (0x79, arg[1], out);
+		return true;
+	} else if (!r_str_casecmp (arg[0], "r2") && resolve_immediate (arg[1] + 1, &src_imm)) {
+		singlearg_immediate (0x7a, arg[1], out);
+		return true;
+	} else if (!r_str_casecmp (arg[0], "r3") && resolve_immediate (arg[1] + 1, &src_imm)) {
+		singlearg_immediate (0x7b, arg[1], out);
+		return true;
+	} else if (!r_str_casecmp (arg[0], "r4") && resolve_immediate (arg[1] + 1, &src_imm)) {
+		singlearg_immediate (0x7c, arg[1], out);
+		return true;
+	} else if (!r_str_casecmp (arg[0], "r5") && resolve_immediate (arg[1] + 1, &src_imm)) {
+		singlearg_immediate (0x7d, arg[1], out);
+		return true;
+	} else if (!r_str_casecmp (arg[0], "r6") && resolve_immediate (arg[1] + 1, &src_imm)) {
+		singlearg_immediate (0x7e, arg[1], out);
+		return true;
+	} else if (!r_str_casecmp (arg[0], "r7") && resolve_immediate (arg[1] + 1, &src_imm)) {
+		singlearg_immediate (0x7f, arg[1], out);
+		return true;
+	} else if (is_indirect_reg (arg[0]) && is_indirect_reg (arg[1])) { //TODO: write doublearg_direct_register for 3 bytes instructions
+		if (!address_direct (arg[0], &dst_addr)) {
+			return false;
 		}
-		return singlearg_direct (0xa8 | register_number (arg[0])
-			, arg[1]
-			, out);
-	}
-	if (!r_str_casecmp (arg[1], "c")) {
-		return singlearg_bit (0x92, arg[0], out);
-	}
-	if (!r_str_casecmp (arg[1], "a")) {
-		return singlearg_direct (0xf5,  arg[0], out);
-	}
-	if (is_reg (arg[1])) {
-		return singlearg_direct (0x88 | register_number (arg[1])
-			, arg[0]
-			, out);
-	}
-	if (is_indirect_reg (arg[1])) {
-		return singlearg_direct (0x86 | register_number (arg[1])
-			, arg[0]
-			, out);
-	}
-	ut8 dest_addr;
-	if (!address_direct (arg[0], &dest_addr)) {
-		return false;
-	}
-	if (arg[1][0] == '#') {
-		ut16 imm;
-		if (!resolve_immediate (arg[1] + 1, &imm)) {
+		if (!address_direct (arg[1], &src_addr)) {
 			return false;
 		}
-		(*out)[0] = 0x75;
-		(*out)[1] = dest_addr;
-		(*out)[2] = imm & 0x00FF;
+		(*out)[0] = 0x85;
+		(*out)[1] = src_addr;
+		(*out)[2] = dst_addr;
 		*out += 3;
+	} else if ((is_indirect_reg (arg[0]) && !r_str_casecmp (arg[1], "@r0")) || (is_indirect_reg (arg[0]) && !r_str_casecmp (arg[1], "[r0]"))) {
+		singlearg_direct (0x86, arg[0], out);
 		return true;
-	}
-	ut8 src_addr;
-	if (!address_direct (arg[1], &src_addr)) {
+	} else if ((is_indirect_reg (arg[0]) && !r_str_casecmp (arg[1], "@r1"))  || (is_indirect_reg (arg[0]) && !r_str_casecmp (arg[1], "[r1]"))) {
+		singlearg_direct (0x87, arg[0], out);
+		return true;
```

Comment:
dont else after return. actually in this case you could just return true once at the end. but the whole code looks very spaguetti and can be probably simplfied a lot theres a lot of copypaste

Reply:
Just for return true or can I esle after a false return?